### PR TITLE
Integrate Rust linter — Clippy, update Rust 1.24.1 → 1.30.0

### PR DESCRIPTION
**Pull request by artem-zinnatullin**
Part of #193.

Had to update Rust and its toolchain to install [Clippy](https://github.com/rust-lang-nursery/rust-clippy), it wasn't part of official toolchain in 1.24.1.

Rust changelog: https://github.com/rust-lang/rust/blob/1.30.0/RELEASES.md

### CODE REVIEW

File: ci/docker/Dockerfile [hunk 1]
```diff
@@ -15,10 +15,18 @@ COPY ci/rust.version /var/rust.version
 # "ant python git": build sample Buck project.
 RUN apt-get update --quiet > /dev/null && \
     apt-get --assume-yes --quiet install sudo openssh-server openjdk-8-jdk \
-    golang clang build-essential lib32stdc++6 lib32z1 unzip ant python git curl && \
-    export MAINFRAMER_RUST_VERSION="$(cat /var/rust.version)" && \
-    curl --silent --fail --location https://static.rust-lang.org/rustup.sh | sh -s -- --revision="$MAINFRAMER_RUST_VERSION" && \
-    rustc --version && cargo --version
+    golang clang build-essential lib32stdc++6 lib32z1 unzip ant python git curl
+
+RUN groupadd build_user && \
```
Comment (artem-zinnatullin) [inline ci/docker/Dockerfile:20]: This step used to happen in entrypoint, but new Rust toolchain doesn't support multiuser installation, I've tried to mess with exposing it for build_user but couldn't make it work.

Thus I add `build_user` at image build time and install Rust toolchain to that user.

Entrypoint just has to make sure that files are owned by proper uid/gid

File: src/args.rs [hunk 2]
```diff
@@ -4,7 +4,7 @@ pub struct Args {
 }
 
 impl Args {
-    pub fn parse(raw_args: Vec<String>) -> Result<Args, String> {
+    pub fn parse(raw_args: &[String]) -> Result<Args, String> {
```
Comment (artem-zinnatullin) [inline src/args.rs:7]:
```console
error: this argument is passed by value, but not consumed in the function body
 --> src/args.rs:7:28
  |
7 |     pub fn parse(raw_args: Vec<String>) -> Result<Args, String> {
  |                            ^^^^^^^^^^^ help: consider changing the type to: `&[String]`
  |
  = note: `-D needless-pass-by-value` implied by `-D warnings`
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_pass_by_value
```

File: src/config.rs [hunk 3]
```diff
@@ -19,7 +19,7 @@ impl Config {
         };
 
         file.read_to_string(&mut content)
-            .expect(&format!("Could not read config file '{}'.", file_path.to_string_lossy()));
+            .unwrap_or_else(|_| panic!("Could not read config file '{}'.", file_path.to_string_lossy()));
```
Comment (artem-zinnatullin) [inline src/config.rs:22]:
```console
error: use of `expect` followed by a function call
  --> src/config.rs:22:14
   |
22 |             .expect(&format!("Could not read config file '{}'.", file_path.to_string_lossy()));
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `unwrap_or_else(|_| panic!("Could not read config file '{}'.", file_path.to_string_lossy()))`
   |
   = note: `-D expect-fun-call` implied by `-D warnings`
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#expect_fun_call
```

File: src/config.rs [hunk 4]
```diff
@@ -34,12 +34,12 @@ fn find_value(config_content: &str, key: &str) -> Option<String> {
         None => None,
         Some(start_index) => {
             let content_starting_with_key = &config_content[start_index..config_content.len()];
-            let value_start_index = match content_starting_with_key.find("=") {
+            let value_start_index = match content_starting_with_key.find('=') {
```
Comment (artem-zinnatullin) [inline src/config.rs:37]:
```console
error: single-character string constant used as pattern
  --> src/config.rs:37:74
   |
37 |             let value_start_index = match content_starting_with_key.find("=") {
   |                                                                          ^^^ help: try using a char instead: `'='`
   |
   = note: `-D single-char-pattern` implied by `-D warnings`
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#single_char_pattern

```

File: src/config.rs [hunk 5]
```diff
@@ -57,7 +57,7 @@ fn find_value(config_content: &str, key: &str) -> Option<String> {
 fn parse_config_from_str(config_content: &str) -> Result<Config, String> {
     Ok(Config {
         remote_machine_name: match find_value(&config_content, &"remote_machine") {
-            None => return Err(format!("please specify 'remote_machine'.")),
+            None => return Err("please specify 'remote_machine'.".to_string()),
```
Comment (artem-zinnatullin) [inline src/config.rs:60]:
```console
error: useless use of `format!`
  --> src/config.rs:60:32
   |
60 |             None => return Err(format!("please specify 'remote_machine'.")),
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using .to_string(): `"please specify 'remote_machine'.".to_string()`
   |
   = note: `-D useless-format` implied by `-D warnings`
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#useless_format

```

File: src/ignore.rs [hunk 6]
```diff
@@ -19,17 +19,20 @@ impl Ignore {
         remote_ignore_file.push(".mainframer/remoteignore");
 
         Ignore {
-            common_ignore_file: match common_ignore_file.exists() {
-                true => Some(common_ignore_file.to_path_buf()),
-                false => None
+            common_ignore_file: if common_ignore_file.exists() {
```
Comment (artem-zinnatullin) [inline src/ignore.rs:22]:
```console
error: you seem to be trying to match on a boolean expression
  --> src/ignore.rs:22:33
   |
22 |               common_ignore_file: match common_ignore_file.exists() {
   |  _________________________________^
23 | |                 true => Some(common_ignore_file.to_path_buf()),
24 | |                 false => None
25 | |             },
   | |_____________^ help: consider using an if/else expression: `if common_ignore_file.exists() { Some(common_ignore_file.to_path_buf()) } else { None }`
   |
   = note: `-D match-bool` implied by `-D warnings`
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#match_bool
```

😠

File: src/sync.rs [hunk 7]
```diff
@@ -59,10 +59,10 @@ fn project_dir_on_remote_machine(local_project_dir_name: &str) -> String {
 
 fn apply_exclude_from(rsync_command: &mut Command, exclude_file: &Option<PathBuf>) {
     match exclude_file {
-        &Some(ref value) => {
+        Some(ref value) => {
```
Comment (artem-zinnatullin) [inline src/sync.rs:62]:
```console
error: you don't need to add `&` to all patterns
  --> src/sync.rs:61:5
   |
61 | /     match exclude_file {
62 | |         &Some(ref value) => {
63 | |             rsync_command.arg(format!("--exclude-from={}", value.to_string_lossy()));
64 | |         }
65 | |         &None => ()
66 | |     };
   | |_____^
```

**Review by artem-zinnatullin — COMMENTED** [review]
Added explanations

File: ci/build.sh [hunk 8]
```diff
@@ -43,7 +43,7 @@ docker build -t mainframer:latest .
 BUILD_COMMAND="set -xe && "
 
 # Setup ssh for tests.
-BUILD_COMMAND+="mkdir -p ~/.ssh/ && "
+BUILD_COMMAND+="whoami && mkdir -p ~/.ssh/ && "
```
Comment (arturdryomov) [inline ci/build.sh]: Do you really need `whoami`?

**Review by arturdryomov — APPROVED** [review]

File: ci/build.sh (hunk 8 above)
Comment (vanniktech) [reply ci/build.sh]: Sometimes Artem is lost and needs to remember who he is.

I'll show myself out. :D

Comment (artem-zinnatullin) [reply ci/build.sh]: This is so true, it hurts lol

**Pull request merged**